api/src/oscc.c

```diff
@@ -492,110 +603,372 @@ oscc_result_t oscc_async_enable( int socket )
+
+              vehicle_detection.has_steering_angle |=
+                  ( rx_frame.can_id == KIA_SOUL_OBD_STEERING_WHEEL_ANGLE_CAN_ID );
+
+              vehicle_detection.has_wheel_speed |=
+                  ( rx_frame.can_id == KIA_SOUL_OBD_WHEEL_SPEED_CAN_ID );
+          case -1:
+              if( EINTR == errno )
+              {
+                  continue;
+              }
+          default:
+              continue;
         }
     }
 
+    close( sock );
 
-    if ( result == OSCC_OK )
+    struct can_contains detection =
     {
-        /* all prior checks will pass even if a valid interface has not been
-           set up - attempt to write an empty CAN frame to the interface to see
-           if it is valid */
-        struct can_frame tx_frame;
+      .is_oscc = oscc_detection.has_brake_report &&
+                 oscc_detection.has_steer_report &&
+                 oscc_detection.has_torqe_report,
+      .has_vehicle = vehicle_detection.has_brake_pressure &&
+                     vehicle_detection.has_steering_angle &&
+                     vehicle_detection.has_wheel_speed
+    };
 
-        memset( &tx_frame, 0, sizeof(tx_frame) );
-        tx_frame.can_id = 0;
-        tx_frame.can_dlc = 8;
+    return detection;
+}
 
-        int bytes_written = write( sock, &tx_frame, sizeof(tx_frame) );
 
-        if ( bytes_written < 0 )
-        {
-            printf( "Failed to write test frame to %s: %s\n", can_channel, strerror(errno) );
+oscc_result_t construct_interfaces_list( struct device_name * const list_ptr )
+{
+    FILE *fh;
+    char buffer[512];
+    oscc_result_t ret = OSCC_ERROR;
 
-            result = OSCC_ERROR;
-        }
-        else
-        {
-            can_socket = sock;
-        }
+    fh = fopen( "/proc/net/dev", "r" );
+    if (!fh) {
+        perror( "Cannot read: /proc/net/dev" );
+
+        return -2;
     }
 
+    //Consume the first two lines since they are headers
+    fgets( buffer, sizeof buffer, fh );
+    fgets( buffer, sizeof buffer, fh );
 
-    return result;
+    while ( fgets( buffer, sizeof buffer, fh ) ) {
+        char *socket_name;
+        socket_name = get_device_name( buffer );
+        ret = add_device_name( socket_name, list_ptr );
+    }
+
+    return ret;
+}
+
+
+char * get_device_name( char *string )
+{
+    size_t span = strcspn(string, ":");
+    static char temp_name[IFNAMSIZ];
+    strncpy(temp_name, string, span);
+    temp_name[span] = '\0';
+    size_t leading_spaces = strspn(temp_name, " ");
+
+    if(leading_spaces == 0)
+    {
+        return temp_name;
+    }
+
+    static char new_name[IFNAMSIZ];
+
+    strncpy(new_name, temp_name + leading_spaces, span - leading_spaces + 1);
+
+    new_name[span - leading_spaces] = '\0';
+
+    return new_name;
+}
+
+oscc_result_t add_device_name( const char * const name,
+                               struct device_name * const list_ptr )
+{
+    if( list_ptr == NULL )
+    {
+        perror( "list is uninitialized" );
```

Comment:
This function is not invoked after a system call that would set `errno`, so this might report bogus errors. In addition if this ever happens then there's a bug in the program. Can we replace this call with an `assert` or some other check that violently terminates the program if the condition is false?